common/src/Microsoft.Azure.IIoT.Hub.Module.Client/src/Default/MqttClient/MqttClientAdapter.cs

```diff
@@ -0,0 +1,616 @@
+            /// <param name="onConnectionLost">Handler for when the MQTT server connection is lost.</param>
+            /// <param name="logger">Logger used for operations</param>
+            /// <returns></returns>
+            public static async Task<IClient> CreateAsync(
+                IManagedMqttClient client,
+                string product,
+                MqttClientConnectionStringBuilder cs,
+                string deviceId,
+                string telemetryTopicTemplate,
+                TimeSpan timeout,
+                IRetryPolicy retry,
+                Action onConnectionLost,
+                ILogger logger) {
+                var options = new MqttClientOptionsBuilder()
+                    .WithClientId(cs.DeviceId)
+                    .WithTcpServer(tcpOptions => {
+                        tcpOptions.Server = cs.HostName;
+                        tcpOptions.Port = cs.Port;
+
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
+                            tcpOptions.BufferSize = 64 * 1024;
+                        }
+                    });
+
+                if (!string.IsNullOrWhiteSpace(cs.Username) && !string.IsNullOrWhiteSpace(cs.Password)) {
+                    options = options.WithCredentials(cs.Username, cs.Password);
+                }
+
+                var tls_options = new MqttClientOptionsBuilderTlsParameters {
+                    UseTls = cs.UsingIoTHub,
+                    SslProtocol = SslProtocols.Tls12,
+                    IgnoreCertificateRevocationErrors = true
+                };
+
+                if (cs.UsingX509Cert) {
+                    tls_options.Certificates = new List<X509Certificate> { cs.X509Cert };
+                }
+                options = options.WithTls(tls_options);
+
+                // Use MQTT 5.0 if message expiry interval is set.
+                if (cs.MessageExpiryInterval != null) {
+                    options = options.WithProtocolVersion(MqttProtocolVersion.V500);
+                }
+
+                var adapter = new MqttClientAdapter(
+                    client,
+                    product,
+                    deviceId,
+                    timeout,
+                    retry,
+                    onConnectionLost,
+                    cs.MessageExpiryInterval,
+                    cs.UsingIoTHub ? MqttQualityOfServiceLevel.AtLeastOnce : MqttQualityOfServiceLevel.ExactlyOnce,
+                    cs.UsingIoTHub,
+                    telemetryTopicTemplate,
+                    logger);
+                client.UseConnectedHandler(adapter.OnConnected);
+                client.UseDisconnectedHandler(adapter.OnDisconnected);
+                client.UseApplicationMessageReceivedHandler(adapter.OnApplicationMessageReceivedHandler);
+
+                var managedOptions = new ManagedMqttClientOptionsBuilder()
+                    .WithClientOptions(options.Build());
+
+                if (cs.UsingStateFile) {
```

Comment:
Please Check if needed, statesync should be handled by Apollo Runtime, if not used for other case please remove from ConnectionString as well as the ManagedMqttClientStorage

Reply:
We should check with cse, seems like this is persisting locally in case a broker is not responding. We don't need this for Apollo integration but maybe it makes sense for standalone and custom in k8s usage